command/agent/cache/lease_cache.go

```diff
@@ -305,7 +308,9 @@ func (c *LeaseCache) Send(ctx context.Context, req *SendRequest) (*SendResponse,
 
 func (c *LeaseCache) createCtxInfo(ctx context.Context) *cachememdb.ContextInfo {
 	if ctx == nil {
+		c.l.RLock()
 		ctx = c.baseCtxInfo.Ctx
+		c.l.RUnlock()
```

Comment:
Do we want to defer in this case since `NewContextInfo` is using that context?

Reply:
Not really required IMHO. We don't want racy access to the `baseCtxInfo` field. After its being read, I think it is okay to use it without the lock. We follow this pattern of releasing the lock as early possible and don't hold up the lock until after the read value is done being used.

Reply:
We only hold the lock if the field in question is being read or set multiple times, which is the other scenario in this PR, where in we are reading the value first and then resetting it.